### PR TITLE
Fix liveshare

### LINKED ISSUE #1625: Jack-in als live share guest not working (polylith repo)
In one of the last versions up to v2.0.257 this stopped working. I don't know which one exactly.

We're using polylith.

When trying to connect to the remote REPL as a live share guest, I'm presented with a list of project roots. The paths are correct, but different from local. Locally, the paths are absolute paths from the root of my file system. Over live share the paths are absolute from the root of the workspace. For example, locally I might have `/Users/my/my-project` and when connecting over live share that would show up as `/`.

I choose ‘/’, then I'm asked for the type, I choose ‘generic’, and then nothing…

### PR DESCRIPTION
**Pull request by svdo**
## What has changed?

Live share support in Calva was broken: as a guest you could no longer connect to the host's REPL. Also using the workspace's `config.edn` file was not handled properly on the guest.

Fixes #1625.

Please note that `npm run eslint` currently fails on the `dev` branch, so it also fails on this branch. My changes don't introduce new linting issues.

## My Calva PR Checklist

If this PR involves code changes, I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
  - [x] Smoke tested the extension as such.
  - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
  - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
  - [ ] ~If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.~
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).
- [ ] ~Add tests or some other mechanism to help prevent breaking changes to live share support in the future~ For now the team is going to try and cover this in their code reviews. Issue #1895 was created to add integration tests for this.

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

### CODE REVIEW

File: src/extension-test/integration/suite/jack-in-test.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   testUtil.log(suite, 'test.clj opened');
 
   const projectRootPath = await projectRoot.findClosestProjectRootPath();
-  const projetcRootUri = vscode.Uri.file(projectRootPath);
+  const projetcRootUri = projectRootPath;
```
Comment (PEZ) [inline src/extension-test/integration/suite/jack-in-test.ts]: Maybe I lack some context, but doesn't it seem like we could do it like so?

``` typescript
const projectRootUri = await projectRoot.findClosestProjectRootPath();
```

File: src/project-root.ts [hunk 2]
```diff
@@ -33,13 +33,13 @@ export function excludePattern(moreExcludes: string[] = []) {
   return `**/{${[...moreExcludes, ...config.getConfig().projectRootsSearchExclude].join(',')}}`;
 }
 
-export async function findClosestProjectRootPath(candidatePaths?: string[]) {
+export async function findClosestProjectRootPath(candidatePaths?: vscode.Uri[]) {
```
Comment (PEZ) [inline src/project-root.ts]: I think all these functions should either be named `...Uri`, or maybe just skip the prefix (if the latter, `findClosestProjectRoot` in this case).

**Comment by PEZ** [comment]
> Please note that `npm run eslint` currently fails on the `dev` branch

It doesn't on my machine. And it also passes in CI...

Seems to be an issue with the integration test though. The Jack-in smoke test fails. Could be a flaky test thing, I am rerunning it now and we'll see.

File: src/project-root.ts (hunk 2 above)
Comment (svdo) [reply src/project-root.ts]: Yes I agree that the names are incorrect. I'll change them.

File: src/extension-test/integration/suite/jack-in-test.ts (hunk 1 above)
Comment (svdo) [reply src/extension-test/integration/suite/jack-in-test.ts]: I think you're right, I'll check & update.

**Comment by svdo** [comment]
> Seems to be an issue with the integration test though. The Jack-in smoke test fails. Could be a flaky test thing, I am rerunning it now and we'll see.

Yes, so cool that there are integration tests though!!! :) 

I'm not sure what's going on here. It also fails on my machine. I guess I'll have to dive into the inner workings of those tests.

**Comment by PEZ** [comment]
The Jack-in test fails like so:

```
Integration testing, Jack-in: waiting for connect...

rejected promise not handled within 1 second: TypeError [ERR_INVALID_ARG_TYPE]: The "options.cwd" property must be of type string or an instance of Buffer or URL. Received an instance of p
stack trace: TypeError: The "options.cwd" property must be of type string or an instance of Buffer or URL. Received an instance of p
	at normalizeSpawnArguments (node:child_process:539:11)
	at Object.spawn (node:child_process:713:13)
	at /home/circleci/calva/out/nrepl/jack-in-terminal.js:66:38
	at new Promise (<anonymous>)
	at JackInTerminal.<anonymous> (/home/circleci/calva/out/nrepl/jack-in-terminal.js:59:20)
	at Generator.next (<anonymous>)
	at /home/circleci/calva/out/nrepl/jack-in-terminal.js:8:71
	at new Promise (<anonymous>)
	at __awaiter (/home/circleci/calva/out/nrepl/jack-in-terminal.js:4:12)
	at JackInTerminal.startClojureProgram (/home/circleci/calva/out/nrepl/jack-in-terminal.js:58:16)
	at JackInTerminal.open (/home/circleci/calva/out/nrepl/jack-in-terminal.js:33:19)
	at c.startSendingEvents (/home/circleci/calva/.vscode-test/vscode-linux-x64-insiders/VSCode-linux-x64/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:93:151182)
```

Seems related to paths vs uris. 😄

**Comment by PEZ** [comment]
> I guess I'll have to dive into the inner workings of those tests.

I think we just need to make sure the `options.cwd` is of the right type. Guessing we are giving it a Uri now.

Of course, Jack-in should fail when you run it manually as well, if it was that simple. I haven't tried this branch yet.

**Comment by svdo** [comment]
Hmm simple jack-in does indeed fail. In my enthusiasm I was only running the live share guest scenario, not the stand-alone scenario. I'll look into it tomorrow, too late for that now.

**Comment by PEZ** [comment]
Thanks @Cyrik for adding the Jack-in step to our integration-test job! ❤️ 🙏

**Comment by svdo** [comment]
Integration tests pass now. Not sure what that "ci/circleci: test-grammar" build error is about?

**Comment by svdo** [comment]
On Slack, @PEZ said:

> Good question about how to prevent these kind of misuse of the API going forward. But I think now that I am aware of it, I can help gate keep. As can the entire [@calva-team](https://clojurians.slack.com/admin/user_groups)

Is that enough to cross the last checkbox and mark this as "ready for review"?

**Comment by PEZ** [comment]
> * [ ] Add tests or some other mechanism to help prevent breaking changes to live share support in the future

It would be great with integration tests validating that certain Live Share scenarios remain supported. However, we could create a separate issue for that to avoid blocking this PR.

**Comment by svdo** [comment]
Thanks @PEZ, I created that issue (#1895). I appreciate that you allow this to speed up things!

File: src/extension-test/integration/suite/jack-in-test.ts [hunk 3]
```diff
@@ -78,8 +78,7 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   await testUtil.openFile(testFilePath);
   testUtil.log(suite, 'test.clj opened');
 
-  const projectRootPath = await projectRoot.findClosestProjectRootPath();
-  const projetcRootUri = vscode.Uri.file(projectRootPath);
+  const projetcRootUri = await projectRoot.findClosestProjectRoot();
```
Comment (bpringe) [inline src/extension-test/integration/suite/jack-in-test.ts]: Would you mind fixing the spelling error? 😄

File: src/file-switcher/file-switcher.ts [hunk 4]
```diff
@@ -54,13 +53,19 @@ export async function toggleBetweenImplAndTest() {
   const newFilename = util.getNewFilename(fileName, extension);
 
   const filePath = path.join(projectRootPath, sourcePath, newFilename);
-  const fileToOpen = vscode.workspace.asRelativePath(filePath, false);
+  let fileToOpen = vscode.workspace.asRelativePath(filePath, false);
+  if (fileToOpen[0] === '/') {
+    // When connected over live share, the above will result in an absolute path,
+    // which causes that workspace.findFiles won't find the file. So we strip the
```
Comment (bpringe) [inline src/file-switcher/file-switcher.ts]:
```suggestion
    // which causes workspace.findFiles to not find the file. So we strip the
```

File: src/file-switcher/file-switcher.ts [hunk 5]
```diff
@@ -54,13 +53,19 @@ export async function toggleBetweenImplAndTest() {
...
+    // leading slash to prevent such scenario's.
```
Comment (bpringe) [inline src/file-switcher/file-switcher.ts]:
```suggestion
    // leading slash to prevent such scenarios.
```

**Review by bpringe — CHANGES_REQUESTED** [review]
Thanks for this PR!

**Comment by bpringe** [comment]
I haven't tested the live share part but if it's working for you @svdo I think this is good to go after the spelling fix.

**Comment by svdo** [comment]
@bpringe Thanks for reviewing and the grammatical improvements. I have tested this with live share as well, though not as much as I'd like because I still cannot use this for work because of the notebook issue. After this is merged I can updated my branch where notebook support is changed and then I can start using this for real. Of course I'll be on top of potential issues that might arise, but I think this is pretty good now.

**Review by bpringe — APPROVED** [review]

**Pull request merged**